multisig-wallet/project/multisig-contract/src/main.sw

```diff
@@ -1,212 +1,141 @@
 dep errors;
 dep events;
+dep interface;
+dep utils;
 
-// Standard library code
-use std::{
-    b512::B512,
-    call_frames::contract_id,
-    constants::ZERO_B256,
-    context::this_balance,
-    ecr::ec_recover_address,
-    hash::sha256,
-    logging::log,
-    token::{
-        force_transfer_to_contract,
-        transfer_to_address,
-    },
-};
-
-use core::num::*;
-
-// Our library imports
-use contract_abi::MultiSignatureWallet;
-use data_structures::{Transaction, User};
+use std::{constants::ZERO_B256, context::this_balance, logging::log, token::transfer};
+
+use data_structures::{SignatureInfo, User};
 use errors::{ExecutionError, InitError};
 use events::{ExecutedEvent, TransferEvent};
+use interface::MultiSignatureWallet;
+use utils::{create_hash, recover_signer};
 
 storage {
-    /// Used to add entropy into hashing of Tx to decrease the probability of collisions / double
-    /// spending
+    /// Used to add entropy into hashing of transaction to decrease the probability of collisions / double
+    /// spending.
     nonce: u64 = 0,
-    /// The number of approvals required in order to execture a Tx
+    /// The number of approvals required in order to execute a transaction.
     threshold: u64 = 0,
-    /// Number of approvals per user
-    weighting: StorageMap<Address, u64> = StorageMap {},
+    /// Number of approvals per user.
+    weighting: StorageMap<b256, u64> = StorageMap {},
 }
 
 impl MultiSignatureWallet for Contract {
-    /// The constructor initializes the necessary values and unlocks further functionlity
-    ///
-    /// # Panics
-    ///
-    /// - When the constructor is called more than once
-    /// - When the user address is the 0th address (0x00000...)
-    /// - When the threshold is set to 0
-    /// - When an owner has an approval weight of 0
     #[storage(read, write)]
-    fn constructor(users: [User; 25], threshold: u64) {
+    fn constructor(threshold: u64, users: Vec<User>) {
         require(storage.nonce == 0, InitError::CannotReinitialize);
         require(threshold != 0, InitError::ThresholdCannotBeZero);
 
         let mut user_index = 0;
-        while user_index < 25 {
-            require(Address::from(ZERO_B256) != users[user_index].identity, InitError::AddressCannotBeZero);
-            require(users[user_index].weight != 0, InitError::WeightingCannotBeZero);
-            storage.weighting.insert(users[user_index].identity, users[user_index].weight);
-            user_index = user_index + 1;
+        while user_index < users.len() {
+            require(ZERO_B256 != users.get(user_index).unwrap().address, InitError::AddressCannotBeZero);
+            require(users.get(user_index).unwrap().weight != 0, InitError::WeightingCannotBeZero);
+
+            storage.weighting.insert(users.get(user_index).unwrap().address, users.get(user_index).unwrap().weight);
+            user_index += 1;
         }
 
         storage.nonce = 1;
         storage.threshold = threshold;
```

Comment:
Should check that threshold <= sum of weights, otherwise wallet can be initialized in an unusable state.